hledger-ui/Hledger/UI/UIState.hs

```diff
@@ -186,10 +188,11 @@ toggleForecast d ui@UIState{aopts=uopts@UIOpts{cliopts_=copts@CliOpts{reportspec
 
 -- | Toggle between showing all and showing only real (non-virtual) items.
 toggleReal :: UIState -> UIState
-toggleReal ui@UIState{aopts=uopts@UIOpts{cliopts_=copts@CliOpts{reportspec_=rspec@ReportSpec{rsOpts=ropts}}}} =
-  ui{aopts=uopts{cliopts_=copts{reportspec_=rspec{rsOpts=toggleReal ropts}}}}
+toggleReal ui@UIState{aopts=uopts@UIOpts{cliopts_=copts@CliOpts{reportspec_=rspec}}} =
+    ui{aopts=uopts{cliopts_=copts{reportspec_=update rspec}}}
   where
-    toggleReal ropts = ropts{real_=not $ real_ ropts}
+    update = either (error "toggleReal: updating Real should not result in an error") id  -- PARTIAL:
```

Comment:
It's unfortunate to have to add more partials when we're trying to minimise those, even though we can convince ourselves they'll never fire. I wonder if we could have a total version of updateReportSpecFromOpts that we could use for safe updates like toggling a flag ?

Reply:
Eg I suppose toggleReal could just perform a custom change to the ReportSpec & ReportOpts without actually regenerating the former ?

Reply:
It could, but this would require climbing through the Query to alter any Real flags. That's tractable, but it wouldn't get around the problem for changing the report periods, which would require a much more complicated algorithm. I just don't see this being a reasonable path.

Reply:
The main problem I see here is that `Query` is a very expressive data type. The CLI generates `Query`s of a very simple form (Dates and flags all at the top, etc.), but this might in the future not be the case. How would we toggle `Real` on the query `Or [And [Acct "a", Real True], Acct "b"]`? It's an un-natural operation. Toggling Real only really makes sense as an option flag, not on the underlying Query.

This gets worse if we start allowing deeply nested depth queries (as I dream might someday be permitted).

Reply:
If the objective is just to get rid of the `error` calls, then we could replace `either (error "…") with `either (const origrspec)`. This is what `setFilter` does. Since the condition will never be fulfilled according to our understanding, it has the same effect as the above code without the partiality.

If the situation changes in the future and this code becomes accessible, it will silently ignore the updates. I'm not sure if this is desirable, or if in a runtime error telling us to fix the code is preferable.